### PR TITLE
File upload filepond

### PR DESCRIPTION
**Pull request by sajberkurajber90**
do not merge !

- if validation exists for certain collection_name - it gets applied. 
- files, which don't pass the validation get deleted.
- on folder change warning modal pops up in case files are in upload queue.
- upload size is set to 1mb just for testing purposes 

open issue:
- on submit the previos collection doesent get reset 
- better error description 
- we need read only folders (i think that topic should be moved to another branch)

please test

### CODE REVIEW

File: resources/js/components/file-pond.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+                    this.isLoadingFiles = this.isLoadingFiles.filter((item) => {
+                        return item !== file.id;
+                    });
+
+                },
+                server: {
+                    process: async (fieldName, file, metadata, load, error, progress, abort, transfer, options) => {
+                        const onSuccess = async (tempFileId) => {
+                            console.log(this.selectedCollection)
```
Comment (patrickweh) [inline resources/js/components/file-pond.js:50]: remove console.log

File: resources/js/components/file-pond.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+                        await $wire.upload('files', file, onSuccess, onError, progress);
+
+                    },
+                    revert: null,
+                    remove: null,
+                }
+            });
+
+            if (typeof lang === 'string' && lang.toLowerCase() === 'de') {
```
Comment (patrickweh) [inline resources/js/components/file-pond.js:71]: not acceptable, no special handling for one specific language.

File: resources/js/components/file-pond.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            }
+            this.pond = create(inputElement, {
+                allowMultiple: true,
+                onaddfilestart: (file) => {
+                    this.isLoadingFiles.push(file.id);
+                },
+                onremovefile: (error, file) => {
+                    if (error) return;
+                    const ids = this.pond.getFiles().map(f => f.serverId);
+                    if (ids.length === 0) {
```
Comment (patrickweh) [inline resources/js/components/file-pond.js:29]: new line before if

File: resources/js/components/file-pond.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            if (!inputElement) {
+                return;
+            }
+            this.pond = create(inputElement, {
+                allowMultiple: true,
+                onaddfilestart: (file) => {
+                    this.isLoadingFiles.push(file.id);
+                },
+                onremovefile: (error, file) => {
+                    if (error) return;
```
Comment (patrickweh) [inline resources/js/components/file-pond.js:27]: new line after if

File: resources/js/components/file-pond.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                },
+                onprocessfile: (error, file) => {
+                    if (error) {
+                        this.pond.removeFile(file.id);
+                    }
+
+                    this.isLoadingFiles = this.isLoadingFiles.filter((item) => {
+                        return item !== file.id;
+                    });
+
```
Comment (patrickweh) [inline resources/js/components/file-pond.js:45]: remove empty line

File: resources/views/livewire/folder-tree.blade.php [hunk 6]
```diff
@@ -1,29 +1,61 @@
...
         selected: false,
         treeSelect(level) {
+            // during file upload, do not allow folder change
+            if(this.isLoadingFiles.length !== 0){
+                return;
+            }
+            // on folder change, clear temp files - if confirmation is accepted
+            if(this.tempFilesId.length !== 0) {
+                window.$wireui.confirmDialog({
+                title: '{{ __('Selected files not submitted') }}',
```
Comment (patrickweh) [inline resources/views/livewire/folder-tree.blade.php:25]: wrong indentation

File: resources/views/livewire/folder-tree.blade.php [hunk 7]
```diff
@@ -1,29 +1,61 @@
...
+                description: '{{ __('Selected files will be deleted on folder change') }}',
+                icon: 'warning',
+                accept: {
+                    label: '{{ __('Confirm') }}',
+                    execute: () => {
+                            this.clearFilesOnLeave();
```
Comment (patrickweh) [inline resources/views/livewire/folder-tree.blade.php:31]: wrong indentation

File: resources/views/livewire/folder-tree.blade.php [hunk 8]
```diff
@@ -194,90 +182,79 @@ function(event) {
...
+                    <x-input :label="__('Size')" disabled x-bind:value="convertSize(selection?.size)"/>
+                    <x-input :label="__('File')" disabled x-bind:value="selection.file_name"/>
+                    <x-input :label="__('Disk')" disabled x-bind:value="selection.disk"/>
                     <x-input x-show="selection?.disk === 'public'"
-                        :label="__('Link')"
-                        readonly
-                        x-ref="originalLink"
-                        type="text"
-                        x-bind:value="selection.original_url"
+                             :label="__('Link')"
```
Comment (patrickweh) [inline resources/views/livewire/folder-tree.blade.php:253]: wrong indentation

File: resources/views/livewire/folder-tree.blade.php [hunk 9]
```diff
@@ -293,21 +270,16 @@ class="h-full rounded-r-md"
                     </x-input>
                 @endcan
                 <object class="object-contain"
-                    x-bind:type="selection.mime_type"
-                    x-bind:data="selection.original_url + '#zoom=85&scrollbar=0&toolbar=0&navpanes=0'"
-                    width="100%"
-                    height="200px"
+                        x-bind:type="selection.mime_type"
+                        x-bind:data="selection.original_url + '#zoom=85&scrollbar=0&toolbar=0&navpanes=0'"
```
Comment (patrickweh) [inline resources/views/livewire/folder-tree.blade.php:274]: wrong indentation

File: resources/views/livewire/folder-tree.blade.php [hunk 10]
```diff
@@ -194,90 +182,79 @@ function(event) {
...
+                              :label="__('Upload')"/>
                 </div>
+
             @endcan
         </div>
         <div x-show="selection.file_name && selected" x-cloak class="flex flex-col gap-3">
             <div class="pb-1.5">
-                <x-button primary :label="__('Download')" x-on:click="$wire.download(selection.id)" />
+                <x-button primary :label="__('Download')" x-on:click="$wire.download(selection.id)"/>
                 @can('action.media.delete')
```
Comment (patrickweh) [inline resources/views/livewire/folder-tree.blade.php:240]: change this to resolve_static(DeleteMedia::class, 'canPerformAction', [false])

File: resources/views/livewire/folder-tree.blade.php [hunk 11]
```diff
@@ -194,90 +182,79 @@ function(event) {
...
-                    <x-button negative :label="__('Delete')" x-on:click="deleteFile(selection)" />
+                    <x-button negative :label="__('Delete')" x-on:click="deleteFile(selection)"/>
                 @endcan
             </div>
             <div class="flex flex-col gap-1.5">
                 @can('action.media.update')
```
Comment (patrickweh) [inline resources/views/livewire/folder-tree.blade.php:245]: change this to resolve_static(UpdateMedia::class, 'canPerformAction', [false])

File: src/Livewire/FolderTree.php [hunk 12]
```diff
@@ -50,10 +52,31 @@ public function loadModel(array $arguments): void
     public function getRules(): array
     {
         return [
-            'files.*' => 'required|file|max:10240',
+            'files.*' => 'required|file|max:1024',
```
Comment (patrickweh) [inline src/Livewire/FolderTree.php:55]: get the max upload file size from livewire config config('livewire.temporary_file_upload.rules')

File: src/Livewire/FolderTree.php [hunk 13]
```diff
@@ -50,10 +52,31 @@ public function loadModel(array $arguments): void
...
+        $modelTypeValidationRules = app($this->modelType)
+            ->query()
+            ->whereKey($this->modelId)
+            ->first()
+            ?->getMediaCollection($baseCollection)
+            ?->acceptsMimeTypes;
+
+        // merge the validation rules
+        return [
+            'files.' . $index => is_null($modelTypeValidationRules) ? 'required|file|max:1024' : 'required|file|max:1024|mimetypes:'
```
Comment (patrickweh) [inline src/Livewire/FolderTree.php:74]: see my comment above

File: tests/Livewire/FolderTreeTest.php [hunk 14]
```diff
@@ -3,13 +3,15 @@
...
 use FluxErp\Livewire\FolderTree;
+use FluxErp\Models\User;
 use FluxErp\Tests\TestCase;
 use Livewire\Livewire;
 
 class FolderTreeTest extends TestCase
 {
     public function test_renders_successfully()
     {
+
```
Comment (patrickweh) [inline tests/Livewire/FolderTreeTest.php:14]: remove this line

File: tests/Livewire/FolderTreeTest.php [hunk 15]
```diff
@@ -3,13 +3,15 @@
 namespace FluxErp\Tests\Livewire;
 
 use FluxErp\Livewire\FolderTree;
+use FluxErp\Models\User;
```
Comment (patrickweh) [inline tests/Livewire/FolderTreeTest.php:6]: unused import

File: resources/js/components/file-pond.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        loadFilePond() {
+            registerPlugin(FilePondPluginImagePreview);
+
+            const inputElement = $ref.querySelector('#filepond-drop');
+            if (!inputElement) {
+                return;
+            }
+            this.pond = create(inputElement, {
+                allowMultiple: true,
```
Comment (patrickweh) [inline resources/js/components/file-pond.js:22]: you should check if the collection accepts multiple or is single file somehow

**Review by patrickweh — CHANGES_REQUESTED** [review]

File: resources/js/components/file-pond.js (hunk 16 above)
Comment (sajberkurajber90) [reply resources/js/components/file-pond.js:22]: if folder containes a single file - and is constrained to one file - i shoud than prevent file upload ?

Comment (patrickweh) [reply resources/js/components/file-pond.js:22]: Warn that the file will be replaced and set allowMultiple to false

Comment (sajberkurajber90) [reply resources/js/components/file-pond.js]: - on uplaod - signgle file folder - file is replaced (I oversaw the watnign window  - will include it tommorow)
- allowMultiple is updated dynamicaly on folder change 
- if folder is single file - the add folder buttton gets hidden.

File: resources/js/components/file-pond.js (hunk 2 above)
Comment (sajberkurajber90) [reply resources/js/components/file-pond.js:80]: i will try async import - to determine which language package should be used for filepond  based on Auth::user()?->language?->language_code -> fallback language is english

Comment (sajberkurajber90) [reply resources/js/components/file-pond.js]: i have added dynamic lang support

File: resources/js/components/file-pond.js (hunk 16 above)
Comment (sajberkurajber90) [reply resources/js/components/file-pond.js]: warning widnow is added - only problem is - Alpine js is throwing the error on rendeLevel(level) -> renderLevel is not defined - although the ui works - will investigate tomorrow

File: resources/js/components/file-pond.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
+import FilePondPluginImagePreview from 'filepond-plugin-image-preview';
+import {create, registerPlugin, setOptions} from 'filepond';
+
+const BASE_LANGUAGE_PATH = '../../../../../node_modules/filepond/locale/';
+
+// load all available languages from filepond
+const availableLanguages = import.meta.glob('../../../../../node_modules/filepond/locale/*.js');
+
+//  TODO: error on tree refresh - renderLevel undefined - and is called several times
```
Comment (SirSplasch) [inline resources/js/components/file-pond.js:9]: is this done, or is this an actual todo?

File: resources/js/components/file-pond.js [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+                },
+                onprocessfile: (error, file) => {
+                    if (error) {
+                        this.pond.removeFile(file.id);
+                    }
+
+                    this.isLoadingFiles = this.isLoadingFiles.filter((item) => {
+                        return item !== file.id;
+                    });
+                    // if single file upload and error is null, show confirm dialog - to replace file
```
Comment (SirSplasch) [inline resources/js/components/file-pond.js:68]: new line before comment

File: src/Livewire/FolderTree.php [hunk 19]
```diff
@@ -92,7 +176,7 @@ public function render(): View|Factory|Application
 
     public function getTree(): array
     {
-        if (! $this->modelType || ! $this->modelId) {
+        if (!$this->modelType || !$this->modelId) {
```
Comment (SirSplasch) [inline src/Livewire/FolderTree.php:179]: revert this change, keep the whitespaces after !

File: src/Livewire/FolderTree.php [hunk 20]
```diff
@@ -21,12 +23,12 @@ class FolderTree extends Component
 {
     use Actions, WithFileUploads;
 
-    /** @var Model $this->modelType */
+    /** @var Model $this ->modelType */
```
Comment (SirSplasch) [inline src/Livewire/FolderTree.php:26]: remove added whitespace between $this and ->

File: src/Livewire/FolderTree.php [hunk 21]
```diff
@@ -21,12 +23,12 @@ class FolderTree extends Component
...
     public ?string $modelType = null;
 
     public ?int $modelId = null;
 
-    public $files = [];
+    public array $files = [];
```
Comment (SirSplasch) [inline src/Livewire/FolderTree.php:31]: are you sure this is always an array? I remember darkly in the back of my mind, that it caused issues setting property type to array.

File: src/Livewire/FolderTree.php [hunk 22]
```diff
@@ -44,16 +46,41 @@ public function loadModel(array $arguments): void
         $this->js(<<<'JS'
             selected = false;
             loadLevels();
-        JS);
+        JS
```
Comment (SirSplasch) [inline src/Livewire/FolderTree.php:49]: > 
        $this->js(
            <<<'JS'
               selected = false;
               loadLevels();
            JS
        );

File: src/Livewire/FolderTree.php [hunk 23]
```diff
@@ -63,20 +90,77 @@ public function updatedFiles(): void
...
+
+        // validation took place in updatedFiles method
+        // so we can safely save the files
         try {
             $media = $this->saveFileUploadsToMediaLibrary(
                 name: 'files',
                 modelId: $this->modelId,
                 modelType: morph_alias($this->modelType),
             );
-
```
Comment (SirSplasch) [inline src/Livewire/FolderTree.php:75]: keep this empty line

**Review by SirSplasch — CHANGES_REQUESTED** [review]

File: resources/js/components/file-pond.js (hunk 17 above)
Comment (sajberkurajber90) [reply resources/js/components/file-pond.js:9]: renderLevel(level) from folder-tree.js - gets triggered several times - onUploadSuccess - need to find solution

File: src/Livewire/FolderTree.php (hunk 21 above)
Comment (sajberkurajber90) [reply src/Livewire/FolderTree.php:31]: will revert

**Review by SirSplasch — APPROVED** [review]

**Review by patrickweh — APPROVED** [review]
